### PR TITLE
Lite: Add_n Op refactored

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
1:> Dynamic memory allocation moved from Eval to Prepare.
2:> VectorOfTensors refactored to eliminate the limitation of 

> "can be used only in Eval"

.

### CODE REVIEW

File: tensorflow/lite/kernels/add_n.cc [hunk 1]
```diff
@@ -51,12 +82,13 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
 
 template <typename T>
 void EvalAddN(TfLiteContext* context, TfLiteNode* node) {
-  // TODO(haoliang): Initialize all_inputs only once during init.
-  VectorOfTensors<T> all_inputs(*context, *node->inputs);
+  OpData* data = reinterpret_cast<OpData*>(node->user_data);
+  VectorOfTensors<T>* all_inputs =
+      static_cast<VectorOfTensors<T>*>(data->all_inputs);
+  int num_inputs = data->num_inputs;
```
Comment (shahzadlone) [inline tensorflow/lite/kernels/add_n.cc]: perhaps ```const```?

File: tensorflow/lite/kernels/internal/tensor.h [hunk 2]
```diff
@@ -72,7 +73,14 @@ class VectorOfTensors {
   // example:
   //   float* const* f = v.data();
   //   f[0][1] is the second element of the first tensor.
-  T* const* data() const { return all_data_.data(); }
+  // NOTE: This function should be called only in Eval/Invoke context
+  T* const* data() {
+    for (std::vector<TfLiteTensor*>::iterator it = all_tensors_.begin();
```
Comment (shahzadlone) [inline tensorflow/lite/kernels/internal/tensor.h]: could```auto``` be used here?

**Comment by ANSHUMAN87** [comment]
@shahzadlone : Your comments are addressed, please check, Thanks!

File: tensorflow/lite/kernels/internal/tensor.h [hunk 3]
```diff
@@ -72,7 +73,13 @@ class VectorOfTensors {
   // example:
   //   float* const* f = v.data();
   //   f[0][1] is the second element of the first tensor.
-  T* const* data() const { return all_data_.data(); }
+  // NOTE: This function should be called only in Eval/Invoke context
+  T* const* data() {
+    for (auto it = all_tensors_.begin(); it != all_tensors_.end(); ++it) {
+      all_data_.push_back(GetTensorData<T>(*it));
```
Comment (haozha111) [inline tensorflow/lite/kernels/internal/tensor.h]: Could you explain a bit about the benefit of refactoring this function?

In your current implementation, I guess if data() is called multiple times, the all_data_ will be appended with same amount of elements again and again? Is that expected behavior?

Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/tensor.h]: > In your current implementation, I guess if data() is called multiple times, the all_data_ will be appended with same amount of elements again and again? Is that expected behavior?

@haozha111 : Sorry i missed that one, Thanks for pointing out, will resolve it.

Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/tensor.h]: > Could you explain a bit about the benefit of refactoring this function?

@haozha111 : The main benefit is, it avoids heap allocation in Eval / Invoke time. Earlier it was allocating every time invoked, now it will be allocated only once, and will be reused every  Eval / Invoke time.

Comment (haozha111) [reply tensorflow/lite/kernels/internal/tensor.h]: Right. Sorry my question is not clear. I mean why do you need change 'VectorOfTensors' class? Does adding a new class member 'all_tensors_' have any benefit compared with before?

Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/tensor.h]: @haozha111 : Yes, earlier 'VectorOfTensors' was reading data directly from Tensors, which will not be available in case of Prepare, so it was a limitation to create instance of 'VectorOfTensors' only in Eval. But now with my change, 'VectorOfTensors'  can be instantiated in Prepare as it stores all the Input Tensors, rather than their data. So when 'VectorOfTensors'->data() is called in Eval, it will fetch all the fresh data loaded into Tensors.

Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/tensor.h]: > > In your current implementation, I guess if data() is called multiple times, the all_data_ will be appended with same amount of elements again and again? Is that expected behavior?
> 
> @haozha111 : Sorry i missed that one, Thanks for pointing out, will resolve it.

This issue is addressed now, please check, Thanks!

File: tensorflow/lite/kernels/add_n.cc [hunk 4]
```diff
@@ -26,6 +26,22 @@ namespace add_n {
...
+
+void* Init(TfLiteContext* context, const char* buffer, size_t length) {
+  // This is a builtin op, so we don't use the contents in 'buffer', if any.
+  // Instead, we allocate a new object to carry information from Prepare() to
+  // Eval().
+  return new OpData();
+}
+
+void Free(TfLiteContext* context, void* buffer) {
+  delete reinterpret_cast<OpData*>(buffer);
```
Comment (haozha111) [inline tensorflow/lite/kernels/add_n.cc]: this frees the OpData, but not the memory pointed by all_inputs.

File: tensorflow/lite/kernels/add_n.cc [hunk 5]
```diff
@@ -42,6 +58,21 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
     TF_LITE_ENSURE_EQ(context, input1->type, input->type);
   }
 
+  OpData* data = reinterpret_cast<OpData*>(node->user_data);
+  data->num_inputs = NumInputs(node);
+  if (output->type == kTfLiteFloat32) {
+    data->all_inputs = reinterpret_cast<void*>(
+        new VectorOfTensors<float>(*context, *node->inputs));
```
Comment (haozha111) [inline tensorflow/lite/kernels/add_n.cc]: Note that you didn't manage to free this piece of memory at the end. It will cause memory leaks.

**Comment by ANSHUMAN87** [comment]
@haozha111 : Thanks a lot for your valuable comments, all are handled now, please check, Thanks!

File: tensorflow/lite/kernels/internal/tensor.h (hunk 3 above)
Comment (haozha111) [reply tensorflow/lite/kernels/internal/tensor.h]: Let's hold on this change a bit. I think this change will make the call a little expensive. The goal of 'VectorOfTensors' should be high efficient. 

Originally VectorOfTensors only handle static tensors (the data won't change), so it's simpler and more efficient. Now as you are supporting this in Prepare stage, where the data of tensor might not be ready. I think we might think a little bit about the trade off here, whether we want to keep the current semantics of the class(efficient, static view of tensors data), or changing its behavior to support dynamic view of tensors data.

Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/tensor.h]: > Let's hold on this change a bit. I think this change will make the call a little expensive. The goal of 'VectorOfTensors' should be high efficient.
> 
> Originally VectorOfTensors only handle static tensors (the data won't change), so it's simpler and more efficient. Now as you are supporting this in Prepare stage, where the data of tensor might not be ready. I think we might think a little bit about the trade off here, whether we want to keep the current semantics of the class(efficient, static view of tensors data), or changing its behavior to support dynamic view of tensors data.

@haozha111 : Sure i will wait for your go ahead. BUt as concern raised by you, just wanted to share my view, if it can help. Actually performance wise, there is no change at all, if you still keep the instantiation in Eval(). But if instantiation kept in Prepare() and data() fetch in Eval(), it actually helps reducing memory allocation every time, performance wise it will still be same or minor improved. Only case where i can see it can go dip is when the data() fetch called multiple times in same sequence where Tensors are not changed at all. As every time it has to refresh and load data. 

Hope my above details helps a bit for your conclusion, Thanks!
Thanks again for your efforts in reviewing my changes!

**Comment by gbaned** [comment]
@haozha111 Could you PTAL and approve.

**Comment by tensorflowbutler** [comment]
Nagging Reviewers @haozha111, @haozha111: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 15 days with no activity and the `awaiting review` label has been applied.

**Comment by gbaned** [comment]
@haozha111, @shahzadlone Can you please review this PR? Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Reviewers @haozha111, @haozha111: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 15 days with no activity and the `awaiting review` label has been applied.

**Comment by ANSHUMAN87** [comment]
@miaout17 , @haozha111 : Gentle Reminder!!!

File: tensorflow/lite/kernels/internal/tensor.h [hunk 6]
```diff
@@ -82,6 +94,7 @@ class VectorOfTensors {
 
  private:
   std::vector<T*> all_data_;
+  std::vector<TfLiteTensor*> all_tensors_;
```
Comment (miaout17) [inline tensorflow/lite/kernels/internal/tensor.h]: Suggestion: remove this field.

File: tensorflow/lite/kernels/internal/tensor.h [hunk 7]
```diff
@@ -72,7 +73,18 @@ class VectorOfTensors {
   // example:
   //   float* const* f = v.data();
   //   f[0][1] is the second element of the first tensor.
-  T* const* data() const { return all_data_.data(); }
+  // NOTE: This function should be called only in Eval/Invoke context
+  T* const* data() {
```
Comment (miaout17) [inline tensorflow/lite/kernels/internal/tensor.h]: Suggestion: Revert this change.

File: tensorflow/lite/kernels/internal/tensor.h [hunk 8]
```diff
@@ -52,12 +52,13 @@ class VectorOfTensors {
     int num_tensors = tensor_list.size;
```
Comment (miaout17) [inline tensorflow/lite/kernels/internal/tensor.h]: Suggestion: Add a `Update` function to update `VectorOfTensors`. It should work just like the constructor.

File: tensorflow/lite/kernels/add_n.cc [hunk 9]
```diff
@@ -26,6 +26,32 @@ namespace add_n {
 constexpr int kInputTensor1 = 0;
 constexpr int kOutputTensor = 0;
 
+struct OpData {
+  int num_inputs;
+  void* all_inputs;
+  TfLiteType type;
```
Comment (miaout17) [inline tensorflow/lite/kernels/add_n.cc]: Suggestion: This can be a `template<typename T> struct OpData` that contains a single `VectorOfTensors<T>`. 

Alternatively, just return `VectorOfTensors<T>`

File: tensorflow/lite/kernels/internal/tensor.h [hunk 10]
```diff
@@ -82,6 +94,7 @@ class VectorOfTensors {
```
Comment (miaout17) [inline tensorflow/lite/kernels/internal/tensor.h]: Only updating data pointer but not updating shapes may be dangerous. The shape can be changed between invocations.

**Review by miaout17 — CHANGES_REQUESTED** [review]
Thanks for the contribution!

The major problem of the change is that the `VectorOfTensors` change seems unsafe, and there should be an easier way to address the TODO. Summarizing here:
* Revert most changes in `VectorOfTensors`
* Add a `Update` function in `VectorOfTensors`
* Call `VectorOfTensors`'s `Update` in Eval

The intention of the TODO is majorly avoiding dynamic memory allocation in `Eval`. This should be sufficient to remove the TODO, and hopefully it will also make the code change simpler.

**Comment by ANSHUMAN87** [comment]
> Thanks for the contribution!
> 
> The major problem of the change is that the `VectorOfTensors` change seems unsafe, and there should be an easier way to address the TODO. Summarizing here:
> 
>     * Revert most changes in `VectorOfTensors`
> 
>     * Add a `Update` function in `VectorOfTensors`
> 
>     * Call `VectorOfTensors`'s `Update` in Eval
> 
> 
> The intention of the TODO is majorly avoiding dynamic memory allocation in `Eval`. This should be sufficient to remove the TODO, and hopefully it will also make the code change simpler.

@miaout17 : Thanks for your effort in reviewing the changes. I have some doubt in your suggestion, please help me clarify. The main idea behind my changes were to store the tensor addresses rather than their data. As the data(and shape) might change, but the tensor will remain same(Please correct me if i am wrong).
So based on your suggestion, if i add update function it will not suffice the TODO resolution.
Your point for shape updates definitely required, i will take care of that.
Also the allocation of all_inputs buffer in Init is required to isolate it from allocation every time in Eval. So it has to be a OPdata Struct as i have done, as we have no knowledge of input data type at that point of time.

Please let me know if i have misunderstood your points. TIA!

File: tensorflow/lite/kernels/add_n.cc (hunk 9 above)
Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/add_n.cc]: @miaout17 : This is required, OpData can not be template based, as we dont know the datat type until prepare is executed. So we can not use in Init or Free, unless we know the datat type.

**Comment by ANSHUMAN87** [comment]
@miaout17 : Thank you for your valuable comments! I have resolved all the comments except one. Please find my comment inline.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by gbaned** [comment]
@miaout17 Can you please take a look on this PR? Thanks!

**Comment by ANSHUMAN87** [comment]
@miaout17 , @gbaned : Gentle Reminder!

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
